### PR TITLE
More-detailed events

### PR DESCRIPTION
**Pull request by paulbellamy**
Fixes: https://github.com/weaveworks/flux/issues/374

Proposed format is:

```
{
  "id": 1234,

  "instanceID": "fuzzy-lunch-14",

  "serviceIDs": ["default/helloworld", "fluxy/fluxsvc"],

  // release|automate|deautomate|lock|unlock
  "type": "release",

  "startedAt": "2017-01-02T01:02:03.456789Z",

  // endedAt will be the same as StartedAt (except for new releases)
  "endedAt": "2017-01-02T01:12:03.456789Z",

  // debug|info|warn|error (for color-coding)
  "logLevel": "info",

  // pre-formated string for errors and other stuff, included for
  // backwards compatibility, but won't be included for new events.
  // Basically, they'll either have "metadata" (below), or a "message".
  "message": "old event message here",

  // event-type specific metadata, might be undefined
  "metadata": {
    "job": { ... }, // Or something else?
    "imageIDs": ["nginx:latest"], // For now there will only be one thing here, but...
    "error": "" // Error string, will be undefined if there wasn't one.
  }
}
```

## TODO

- [x] Figure out how to include the release spec and result in the format
  - [x] Integrate changes from https://github.com/weaveworks/flux/pull/421 for new releaseresult stuff
- [x] Ask simon if the format works for him
- [x] Check this doesn't break the UI (or fix the UI at the same time)
- [x] Double-check backwards-compatibility
- [ ] ~~Do fluxctl release when merging this~~
- [ ] Drop the old history table (https://github.com/weaveworks/flux/pull/470)

### CODE REVIEW

**Comment by paulbellamy** [comment]
@foot what is your preference for how to deploy this with the UI. Existing history format is:

```json
{ "Stamp": "2017-06-01T01:02:03.456Z",
  "Type": "v0",
  "Data": "some description string here" }
```

I'm still generating a "Data" description, so I can just increment the "Type" field on all events and begin including new fields. That means the UI will need to support the new fields first...

**Comment by paulbellamy** [comment]
@foot, Ok format of the history events with this is:
```json
[
  {
    "Data": "Released: no image changes to default/helloworld, flux-example/nginx",
    "Event": {
      "endedAt": "2017-02-20T10:18:32.509249Z",
      "id": 1,
      "logLevel": "info",
      "message": "Starting",
      "serviceIDs": [
        "default/helloworld"
      ],
      "startedAt": "2017-02-20T10:18:32.509249Z",
      "type": "release"
    },
    "Stamp": "2017-02-21T12:50:09.086339Z",
    "Type": "v0"
  }
]
```

Am happy to add more/less fields if it makes the UI bit you want to do easier.

**Comment by foot** [comment]
Beaut.

**Comment by foot** [comment]
Looking at the most recent events coming out of dev the new summaries look like this 

> ui-server: Release ui-server (quay.io/weaveworks/ui-server:master-ec5e5e3 -> master-d38e7fe) succeeded

Recording
 - `previousImage` (Either `{tag, host, path}` struct or string).
 - `nextImage`
 - `status` (success|failure)

Which would be great to have at some point. They also appear to have been split up a bit so there's a single event per service (which is fine, just noting it, and perhaps `serviceIDs` -> `serviceID` if thats the new thing.)

**Comment by paulbellamy** [comment]
Ok, there might be multiple previous and next images. 

> They also appear to have been split up a bit so there's a single event per service (which is fine, just noting it, and perhaps serviceIDs -> serviceID if thats the new thing.)

Other way around I've combined the events so there is one per release, which may involve multiple image ids, and multiple services.

How about:

```json
[
  {
    "Data": "Released: quay.io/weaveworks/helloworld:master-a000001 to default/helloworld",
    "Event": {
      "endedAt": "2017-02-22T13:51:17.401383Z",
      "id": 11,
      "logLevel": "error",
      "metadata": {
        "error": "sending notifications: 404 NOT FOUND from Slack (Not found)",
        "release": {
          "createdAt": "2017-02-22T13:51:08.553378Z",
          "done": true,
          "endedAt": "2017-02-22T13:51:17.401383182Z",
          "id": "4cb38145-63a6-9497-c284-d27ffd181e0d",
          "instanceID": "1",
          "log": [
            "Queued.",
            "Calculating updates for release.",
            "Cloning git repository.",
            "Finding defined services.",
            "Found service default/helloworld",
            "Found 1 services.",
            "Looking up images.",
            "Will update default/helloworld container helloworld: quay.io/weaveworks/helloworld:master-a000002 -> master-a000001",
            "Pushing changes.",
            "Applying changes."
          ],
          "priority": 200,
          "result": {
            "default/helloworld": {
              "Error": "",
              "PerContainer": [
                {
                  "Container": "helloworld",
                  "Current": "quay.io/weaveworks/helloworld:master-a000002",
                  "Target": "quay.io/weaveworks/helloworld:master-a000001"
                }
              ],
              "Status": "success"
            }
          },
          "spec": {
            "Excludes": null,
            "ImageSpec": "quay.io/weaveworks/helloworld:master-a000001",
            "Kind": "execute",
            "ServiceSpec": "",
            "ServiceSpecs": [
              "default/helloworld"
            ]
          },
          "startedAt": "0001-01-01T00:00:00Z",
          "status": "success"
        }
      },
      "serviceIDs": [
        "default/helloworld"
      ],
      "startedAt": "2017-02-22T13:51:17.601052Z",
      "type": "release"
    },
    "Stamp": "2017-02-22T13:51:17.601052Z",
    "Type": "v0"
  }
]
```

Metadata is specific to the type of event, and is only set for "release"-type events.

There is a `Event.metadata.error"` field. it will not be present if the whole release was successful.

Then in `Event.metadata.release.result`, there is an entry for each service we attempted to update, with the error, status, and image updates (might be multiple).

Then there is `Event.metadata.release.spec` which is what the user requested to happen. The actual input to the release. Please ignore `ServiceSpec` (deprecated), and use `ServiceSpecs`.

There is also a `Event.serviceIDs` at the top-level, it should agree with the ones in `result`, but will be set regardless of the `Event.type`

File: server/server.go [hunk 1]
```diff
@@ -236,9 +234,10 @@ func (s *Server) History(inst flux.InstanceID, spec flux.ServiceSpec) (res []flu
 	res = make([]flux.HistoryEntry, len(events))
 	for i, event := range events {
 		res[i] = flux.HistoryEntry{
-			Stamp: &events[i].Stamp,
+			Stamp: &events[i].StartedAt,
 			Type:  "v0",
-			Data:  fmt.Sprintf("%s: %s", event.Service, event.Msg),
+			Data:  event.String(),
```
Comment (squaremo) [inline server/server.go]: Oh I see, it's used here. OK, maybe not a big deal then.

File: docker/Dockerfile.flux-db [hunk 2]
```diff
@@ -0,0 +1,3 @@
+FROM postgres:9.4
+
+ADD create.sql /docker-entrypoint-initdb.d/
```
Comment (paulbellamy) [inline docker/Dockerfile.flux-db]: eh? This file shouldn't exist anymore...

File: event.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+type EventID int64
+
+type Event struct {
+	// ID is a UUID for this event. Will be auto-set when saving if blank.
+	ID EventID `json:"id"`
+
+	// InstanceID affected by this event. Note, this is not exposed to the
+	// public.
+	InstanceID InstanceID `json:"-"`
```
Comment (paulbellamy) [inline event.go]: I didn't consider that. Will have a look.

File: history/sql/sql.go [hunk 4]
```diff
@@ -1,86 +1,82 @@
...
-		var event history.Event
-		eventRows.Scan(&event.Service, &event.Msg, &event.Stamp)
-		events = append(events, event)
-	}
+	var events []flux.Event
+	for rows.Next() {
+		var (
+			h             flux.Event
+			instanceID    string
+			serviceIDs    pq.StringArray
```
Comment (paulbellamy) [inline history/sql/sql.go]: Works fine. It's a total hack, though. Will clean it up.

File: release/releaser.go [hunk 5]
```diff
@@ -153,20 +153,83 @@ func (r *Releaser) release(instanceID flux.InstanceID, job *jobs.Job, logStatus
...
 	applyErr := applyChanges(rc.Instance, updates, results)
 	timer.ObserveDuration()
+
+	status := flux.ReleaseStatusSuccess
+	if applyErr != nil {
+		status = flux.ReleaseStatusFailed
+	}
+	release := flux.Release{
+		ID:         flux.ReleaseID(job.ID),
+		InstanceID: job.Instance,
```
Comment (squaremo) [inline release/releaser.go]: Is this still needed?

File: release.go [hunk 6]
```diff
@@ -45,14 +45,15 @@ func NewReleaseID() ReleaseID {
...
-	ID        ReleaseID            `json:"id"`
-	CreatedAt time.Time            `json:"createdAt"`
-	StartedAt time.Time            `json:"startedAt"`
-	EndedAt   time.Time            `json:"endedAt"`
-	Done      bool                 `json:"done"`
-	Priority  int                  `json:"priority"`
-	Status    ServiceReleaseStatus `json:"status"`
-	Log       []string             `json:"log"`
+	ID         ReleaseID            `json:"id"`
+	InstanceID InstanceID           `json:"instanceID"`
```
Comment (squaremo) [inline release.go]: Still needed? I think it was here so it could be fished out and used to log the event, but that doesn't happen any more.

**Pull request merged**